ArjunaJTA/jta/classes/com/arjuna/ats/jta/common/JTAEnvironmentBean.java

```diff
@@ -1225,4 +1236,127 @@ public void setNotifyCommitMarkableResourceRecoveryModuleOfCompleteBranches(
+     */
+    public void setSubordinateTransactionImporterClassName(String importerClassName) {
+        synchronized(this)
+        {
+            if(importerClassName == null)
+            {
+                this.subordinateTransactionImporter = null;
+            }
+            else if(!importerClassName.equals(this.subordinateTransactionImporterClassName))
+            {
+                this.subordinateTransactionImporter = null;
+            }
+            this.subordinateTransactionImporterClassName = importerClassName;
+        }
+    }
+
+    public String getSubordinateTransactionImporterClassName(){
+        return this.subordinateTransactionImporterClassName;
+    }
+
+    public void setSubordinateTransactionImporter(SubordinateTransactionImporter importer) {
+        synchronized(this) {
+            SubordinateTransactionImporter oldImporter = this.subordinateTransactionImporter;
+            this.subordinateTransactionImporter = importer;
+
+            if(importer == null)
+            {
+                this.subordinateTransactionImporterClassName = null;
+            }
+            else if(importer != oldImporter)
+            {
+                this.lastResourceOptimisationInterfaceClassName = ClassloadingUtility.getNameForClass(importer);
+            }
+            this.subordinateTransactionImporter = importer;
+        }
+    }
+    
+    /**
+     * Importing subordinate transaction.
+     */
+    public SubordinateTransactionImporter getSubordinateTransactionImporter() {
+        if(subordinateTransactionImporter == null && subordinateTransactionImporterClassName != null) {
+            synchronized(this) {
+                if(subordinateTransactionImporter == null && subordinateTransactionImporterClassName != null) {
+                    subordinateTransactionImporter = ClassloadingUtility.loadAndInstantiateClass(SubordinateTransactionImporter.class,
+                        subordinateTransactionImporterClassName, null);
+                }
+            }
+        }
+
+        return subordinateTransactionImporter;
+    }
+
+    /**
+     * Setting of class name that defines {@link UserTransactionOperationsProvider}.
+     * The provider is later used to get more info about {@link UserTransaction}.
+     *
+     * @param providerClassName  class name implementing {@link UserTransactionOperationsProvider}
+     */
+    public void setUserTransactionOperationsProviderClassName(String providerClassName) {
+        synchronized (this) {
+            if(providerClassName == null)
+            {
+                this.userTransactionOperationsProvider = null;
+            }
+            else if(!providerClassName.equals(this.userTransactionOperationsProviderClassName))
+            {
+                this.userTransactionOperationsProvider = null;
+            }
+            this.userTransactionOperationsProviderClassName = providerClassName;
+        }
```

Comment:
Again I took it as a pattern used whole over the class. For this could be fixed at your way I need to change tests as they depend on current behavior - when null is set then is expected to get null. If you ask me why it's so I have no idea ;o)